app/uk/gov/hmrc/apiplatformmicroservice/metrics/RecordMetrics.scala

```diff
@@ -0,0 +1,37 @@
+/*
+ * Copyright 2020 HM Revenue & Customs
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package uk.gov.hmrc.apiplatformmicroservice.metrics
+
+import scala.concurrent.{ExecutionContext, Future}
+import scala.util.{Failure, Success}
+
+trait RecordMetrics {
+  val apiMetrics: ApiMetrics
+  val api: API
+
+  def record[A](f: => Future[A])(implicit ec: ExecutionContext): Future[A] = {
+    val timer = apiMetrics.startTimer(api)
+
+    f.andThen {
+        case _ => timer.stop()
+      }
+      .andThen {
+        case Success(_) => apiMetrics.recordSuccess(api)
+        case Failure(_) => apiMetrics.recordFailure(api)
+      }
```

Comment:
just wonder if we should have a recover/catch clause because not all errors get caught in the future.failed (though most/almost all do)

Reply:
What errors wouldn't be caught in there? The method takes in a lazy future. Wouldn't all failed futures fall into the failure case?

Reply:
Surprisingly f can throw, rather than returning a failed future.  That's not to say that any instance of f in this could base right now does, just that it is possible given the signature.

Reply:
I'm not so worried that you should hold up the changes though.

Reply:
Ah I see what you mean. I've made a change and now it also records failures when f throws an exception.